### DISCUSSION TITLE
Program with shader crashes if shader modifies uniform vec4. Is this a bug?

### DISCUSSION THREAD

**Discussion by imprity** [hajimehoshi/ebiten#2842] [Q&A] [answered]
While I was writing a shader code, I noticed that the program crashes when it compiles the shader that modifies the uniform vec4 component.

I thought about making a bug report but I'm very new to shader programming and wasn't sure if it is a bug.
So I'm writing it here.

But even if it's not a bug. I think it would be helpful if it's written in documentation somewhere.

Here's the minimum code I came up with that crashes when I compile and run.

```
package main

import (
	"github.com/hajimehoshi/ebiten/v2"
	"log"
)

const (
	ScreenWidth  = 600
	ScreenHeight = 600
)

const ShaderCode = "" +
"package main\n" +
"var UniformVec4 vec4\n" +
"\n" +
"func Fragment(position vec4, texCoord vec2, color vec4) vec4{\n" +
"	UniformVec4.g *= 0.5\n" + //comment this line out to make it run
"	return UniformVec4\n" +
"}"

var Shader *ebiten.Shader

type Game struct {}

func (g *Game) Update() error {
	return nil
}

func (g *Game) Draw(screen *ebiten.Image) {
	shaderOp := &ebiten.DrawRectShaderOptions{}

	shaderOp.Uniforms = map[string]any{
		"UniformVec4" : []float32{0.0, 1.0, 0.0, 1.0},
	}

	screen.DrawRectShader(ScreenWidth, ScreenHeight, Shader, shaderOp)
}

func (g *Game) Layout(outsideWidth, outsideHeight int) (int, int) {
	return ScreenWidth, ScreenHeight
}

func main() {
	var err error

	if Shader, err =  ebiten.NewShader([]byte(ShaderCode)); err != nil{
		log.Fatal(err)
	}

	ebiten.SetWindowSize(ScreenWidth, ScreenHeight)
	ebiten.SetWindowTitle("Cool Window")
	if err := ebiten.RunGame(&Game{}); err != nil {
		log.Fatal(err)
	}
}
```

If someone else is having similar issues, I was able to solve it by making a copy.
`copy := UniformVec4`

**Comment by hajimehoshi** [comment]
What crash did you see?

**Reply by hajimehoshi** [reply, marked as answer]
Yeah this is expected. Uniforms variables cannot be modified. This is not a bug.

**Reply by imprity** [reply]
Thanks for the reply, and for the awesome software. I checked the documents again and it does say you shouldn't modify the uniform variables. And now I feel super dumb. Sorry.
